src/Workspaces/Core/Portable/Workspace/Solution/Solution.CompilationTracker.cs

```diff
@@ -863,6 +863,41 @@ public IEnumerable<SyntaxTree> GetSyntaxTreesWithNameFromDeclarationOnlyCompilat
                 return version;
             }
             #endregion
+
+            private class MetadataReferenceComparer : IEqualityComparer<MetadataReference>
+            {
+                public static readonly MetadataReferenceComparer Instance = new MetadataReferenceComparer();
+
+                public bool Equals(MetadataReference x, MetadataReference y)
+                {
+                    if (object.ReferenceEquals(x, y))
+                    {
+                        return true;
+                    }
+
+                    // here we only special case compilation reference and not portable executable reference since
+                    // consumer of this make sure they share metadata reference between compilations.
+                    var compilation1 = x as CompilationReference;
+                    var compilation2 = y as CompilationReference;
+                    if (compilation1 == null || compilation2 == null)
+                    {
+                        return false;
+                    }
+
+                    return object.ReferenceEquals(compilation1.Compilation, compilation2.Compilation);
```

Comment:
This isn't checking the reference properties (i.e. aliasing, embedded interop types.) I suspect you need to.